src/main/kotlin/org/rust/ide/injected/RsDoctestLanguageInjector.kt

```diff
@@ -0,0 +1,176 @@
+    }
+
+    override fun elementsToInjectIn(): List<Class<out PsiElement>> =
+        listOf(RsDocCommentImpl::class.java)
+
+    override fun getLanguagesToInject(registrar: MultiHostRegistrar, context: PsiElement) {
+        if (context !is RsDocCommentImpl) return
+        if (!context.isValidHost || context.elementType !in RS_DOC_COMMENTS) return
+        if (!context.project.rustSettings.doctestInjectionEnabled) return
+
+        val rsElement = context.ancestorStrict<RsElement>() ?: return
+        val cargoTarget = rsElement.containingCargoTarget ?: return
+        if (!cargoTarget.isLib) return // only library targets can have doctests
+        val crateName = cargoTarget.normName
+        val text = context.text
+
+        findDoctestInjectableRanges(text, context.elementType).map { ranges1 ->
+            val ranges = ranges1.map {
+                CodeRange(
+                    it.startOffset,
+                    it.endOffset,
+                    CharArrayUtil.shiftForward(text, it.startOffset, it.endOffset, " \t")
+                )
+            }
+            val commonIndent = ranges.filter { it.isCodeNotEmpty() }.map { it.indent }.min()
+            val indentedRanges = if (commonIndent != null) ranges.mapNotNull { it.offsetIndent(commonIndent) } else ranges
+
+            indentedRanges.map { (start, end, codeStart) ->
+                // `cargo doc` has special rules for code lines which start with `#`:
+                //   * `# ` prefix is used to mark lines that should be skipped in rendered documentation.
+                //   * `##` prefix is converted to `#`
+                // See https://github.com/rust-lang/rust/blob/5182cc1c/src/librustdoc/html/markdown.rs#L114
+                when {
+                    text.startsWith("##", codeStart) -> TextRange(codeStart + 1, end)
+                    text.startsWith("# ", codeStart) -> TextRange(codeStart + 2, end)
+                    else -> TextRange(start, end)
+                }
+            }
+        }.forEach { ranges ->
+            val inj = registrar.startInjecting(RsLanguage)
+
+            ranges.forEachIndexed { index, range ->
+                val isFirst = index == 0
+                val isLast = index == ranges.size - 1
+
+                val prefix = if (isFirst) {
+                    buildString {
+                        if (crateName != "std" || cargoTarget.pkg.origin != PackageOrigin.STDLIB) {
+                            append("extern crate ")
+                            append(crateName)
+                            append("; ")
+                        }
+                        append("fn main() {")
+                    }
+                } else {
+                    null
+                }
+                val suffix = if (isLast) "}" else null
+
+                inj.addPlace(prefix, suffix, context, range)
+            }
+
+            inj.doneInjecting()
+        }
+    }
```

Comment:
Maybe it worth to extract lambda from `forEach` and `map` into separate functions?
Looks like it can simplify reading

Reply:
Too lot of local variables are used there (